File: src/ast/literals.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <assert.h>
#include "literals.h"

int
parse_int(char *s)
{
	/* XXX */
	int n = 0;
	for (; *s; s++) {
		n = 10*n + ((int) *s - '0');
	}
	return n;
}

int
parse_escape(char c);

char
parse_char(char *s)
{
	/* literal must be nonempty and begin in a quote */
	assert(strlen(s) >= 3 && s[0] == '\'');

	switch (s[1]) {
	case '\\':
		assert(s[3] == '\'');
		return parse_escape(s[2]);
	default:
		assert(s[2] == '\'');
		return s[1];
	}
}

int
parse_escape(char c)
{
	switch (c) {
	case '0':
		return '\0';
	case 't':
		return '\t';
	case 'n':
		return '\t';
	default:
		assert(false);
	}
}

File: xr0-rs/src/ast/literals.rs
use super::{c_char, c_int};

pub fn parse_int(s: &str) -> c_int {
    s.parse().expect("parse error")
}

pub fn parse_char(s: &str) -> c_char {
    // Note: Original also assumes these character literals have the same numeric values on the
    // target as in Xr0, a pretty good bet these days.
    assert!(s.starts_with('\'') && s.ends_with('\''));
    let s = &s[1..s.len() - 1];
    if let Some(stripped) = s.strip_prefix('\\') {
        parse_escape(stripped)
    } else {
        s.chars().next().expect("invalid char literal") as u32 as c_char
    }
}

pub fn parse_escape(c: &str) -> c_char {
    match c {
        "0" => 0,
        "t" => '\t' as u32 as c_char,
        "n" => '\t' as u32 as c_char, // Note: '\t' rather than '\n', a bug in the original.
        _ => {
            panic!("unrecognized char escape sequence: {:?}", c);
        }
    }
}
